### PR TITLE
Clarify continuity.

### LINKED ISSUE #3502: Can der apply to a discrete Real variable
``` modelica
model DiscreteRealDerTest
  Real xd;
  Real x;
  Real y;
  // Real yd = der(xd); // This is not allowed according to MLS3.6 P44
equation
  when sample(0, 0.1) then
    xd = sin(time);
  end when;
  x = if time < 0.5 then sin(time) else xd;
  y = der(x); // Should this be supported? Should a warning message given if supported?
end DiscreteRealDerTest;
```

In the above model, `Real yd = der(xd)` should not allowed  according to MLS3.6 P44 since `xd` is discrete-time Real variable, see bottom.  `x` is continue-time Real variable, so it could be derivatived, however `der(x) = if time < 0.5 then cos(time) else der(xd);` involed in `der(xd)` which is not allowed.

In OpenModelica lastest release v1.22.3, 

1. `DiscreteRealDerTest` can be simulated, and `y = if time < 0.5 then cos(time) else 0;`
2.  `Real yd = der(xd);`  will get a compile error.

`DiscreteRealDerTest` is a simplify model of `Modelica.Electrical.Spice3.Examples.Nor` and `Modelica.Electrical.Spice3.Examples.Spice3BenchmarkFourBitBinaryAdder` in Modelica Standard Library 4.0, this two standard models both use `Modelica.Electrical.Spice3.Sources.V_pulse` which resulted in derivative a discrete variable.

image

### PR DESCRIPTION
**Pull request by HansOlsson**
Closes #3502
Just adding an example.

### CODE REVIEW

File: chapters/operatorsandexpressions.tex [hunk 1]
```diff
@@ -782,6 +782,27 @@ \subsection{Derivative and Special Purpose Operators with Function Syntax}\label
...
 For \lstinline!Real! parameters and constants the result is a zero scalar or array of the same size as the variable.
 \end{nonnormative}
+
+\begin{example}
+For continuous expression we have several cases.
+\begin{lstlisting}[language=modelica]
+  Real when1, x1, x2, x3, w1, y1, y2, y3;
+equation
+  when sample(1, 1) then
+    when1 = ...;
```
Comment (henrikt-ma) [inline chapters/operatorsandexpressions.tex:792]: Use math:
```suggestion
    when1 = $\ldots$;
```

**Review by henrikt-ma — COMMENTED** [review]
The idea that one should dynamically check for discontinuities is new to me.  I will need to test implement this in order complete the review.  I also encourage other tool vendors who are new to the idea of runtime checking to test implement and report results.

To clarify the need for test implementation: I am afraid that there could be many models which are considered working fine despite discontinuities which are considered normal at events despite `der` becoming ill-defined at these points in time.

It will be interesting to see if `der` being applied to discontinuous trajectories is concentrated to certain libraries…

**Comment by HansOlsson** [comment]
> The idea that one should dynamically check for discontinuities is new to me. I will need to test implement this in order complete the review. I also encourage other tool vendors who are new to the idea of runtime checking to test implement and report results.
> 
> To clarify the need for test implementation: I am afraid that there could be many models which are considered working fine despite discontinuities which are considered normal at events despite `der` becoming ill-defined at these points in time.

Well, since potential discontinuities may involve arbitrary expressions and we don't have a truth-solver for that we would have to do one of the following:
- Generate warnings (or errors) for things that are in fact continuous. (Hint: Dymola did that earlier, people weren't happy.)
- Allow discontinuities when it is too complicated, breaking the underlying assumptions.

**Comment by henrikt-ma** [comment]
> * Allow discontinuities when it is too complicated, breaking the underlying assumptions.

The problem is that I am not convinced that users (library authors included) generally assume operands of `der` to be differentiable at events.

**Comment by HansOlsson** [comment]
> > * Allow discontinuities when it is too complicated, breaking the underlying assumptions.
> 
> The problem is that I am not convinced that users (library authors included) generally assume operands of `der` to be differentiable at events.

Unless impulses are generated (which is a complicated issue) the results will be incorrect, and people normally want correct results.

**Comment by henrikt-ma** [comment]
> Unless impulses are generated (which is a complicated issue) the results will be incorrect, and people normally want correct results.

They want results to be correct with respect to some specification, and I am not sure this specification has to be one where the differential equations are fulfilled at events.  The details of how to leave enough room for system reinitialization without opening up for arbitrary results I don't know, but to me this approach seems the most promising for one day getting a proper definition of `reinit` semantics.

**Review by eshmoylova — COMMENTED** [review]
Reviewed and forwarded to the team that deals with differentiation. Waiting for their response.

**Review by eshmoylova — APPROVED** [review]

**Comment by HansOlsson** [comment]
Could mention exception for reinit.

**Comment by HansOlsson** [comment]
Poll:
1. Add as is
2. Add exception for reinit (exclude at all events there is a reinit)
3. Delay PR until additional test-implementations

Henrik (3), Elena (1 & 2), Markus(abstain), Gerd (2 maybe 3), Dag (abstain), Hans (1 & 2).
2 slightly ahead, I think

**Pull request merged**